Parse TOML to a nested struct

SouthEndMusic opened the pull request:
Fixes https://github.com/Deltares/Wflow.jl/issues/243
Fixes https://github.com/Deltares/Wflow.jl/issues/688

The original idea was to use `Configuration.jl` to parse the TOML, but after discussion with @JoostBuitink and @vers-w we decided not to do this. Instead the implementation is custom, using `AbstractConfigSection` and a parse function `init_config_section`. 

An overview of the changes made in this PR:
- The package `PropertyDicts` is now used to access nested dictionaries as if they were nested structs (not sure if this is still actually required though)
- Conversion to the proper type of a config field is done in `init_config_section`, so the types do not have to be specified anymore in the code where the config field is used
- The package `EnumX` is used to make enumerations of config options, see the top of `config_structure.jl`. These enumerations are also used to validate these config options
- Some work was done to write the `Config` object back to `TOML` with the function `to_dict`, which is a requirement for https://github.com/Deltares/Wflow.jl/issues/701
- Basic printing functionality was added for the `Config` object mainly for developers, but this could be added to the log if desirable
- Some constructors were removed that are now automatically generated with the `@with_kw` macro
- Some function names were changed to reflect that they are constructors of a particular type
- Some function signatures were made more specific for clarity

vers-w commented on Wflow/src/network.jl:
```diff
@@ -291,11 +289,11 @@ function NetworkReservoir(dataset::NCDataset, config::Config, network::NetworkRi
     # note that these locations are only the reservoir outlet pixels
-    lens = lens_input(config, "reservoir_location__count"; optional = false)
+    lens = lens_input(:reservoir_location__count)
     locs = ncread(dataset, config, lens; sel = indices, type = Int, fill = 0, logging)
 
     # this holds the same ids as locs, but covers the entire reservoir
-    lens = lens_input(config, "reservoir_area__count"; optional = false)
+    lens = lens_input(:reservoir_area__count)
```
Should `config` not be added as an argument, to check if these keys are included, or is this done somewhere else?

SouthEndMusic replied:
`haskey` does not work on the fields that are hardcoded in `config.jl`. If the field is (universally) mandatory but omitted in the TOML, this yields an error in `Configurations.jl`.

vers-w commented on Wflow/src/routing/reservoir.jl:
```diff
@@ -61,7 +61,8 @@ function ReservoirParameters(dataset::NCDataset, config::Config, network::Networ
 
-    lens = lens_input(config, "reservoir~lower_location__count")
+    # TODO: I don't understand what this is supposed to read
+    lens = (; name = "reservoir~lower_location__count", lens = nothing)
```
This is reading the reservoir location ID of a lower (downstream) reservoir when reservoirs are linked. This is possible for reservoirs with a rating curve type (`ouflowfunc`) 2.

SouthEndMusic replied:
There's no test for this, maybe a follow-up issue is required?

vers-w replied:
There is indeed no integrated/model test for this, there is only a unit test for linked reservoirs. This optional key is read from the TOML file:

```toml
[input]
"reservoir~lower_location__count" = "linked_reservoirs"
```
The current implementation (this PR) now sets this parameter always to its default value, that is not correct.

SouthEndMusic replied:
I think I'll handle this as suggested [here](https://github.com/Deltares/Wflow.jl/pull/689#discussion_r2336608541), because tildes are not allowed in field names

vers-w commented on Wflow/src/routing/reservoir.jl:
```diff
@@ -73,7 +74,7 @@ function ReservoirParameters(dataset::NCDataset, config::Config, network::Networ
 
     n_reservoirs = length(area)
-    lens = lens_input(config, "reservoir_location__count"; optional = false)
+    lens = lens_input(:reservoir_location__count)
```
Is `config` argument not required here to check if key is included, or is this handled somewhere else in the code?

SouthEndMusic replied:
See the similar comment

vers-w commented on Wflow/src/routing/surface_kinwave.jl:
```diff
@@ -17,14 +17,15 @@ function init_kinematic_wave_timestepping(
         timestepping = TimeStepping(; stable_timesteps, adaptive)
     else
-        dt_fixed = get(config.model, "$(domain)_kinematic_wave__time_step", dt_fixed)
+        dt_config = getfield(config.model, Symbol("$(domain)_kinematic_wave__time_step"))
+        dt_fixed = isnothing(dt_config) ? dt_fixed : dt_config
```
This can be set from the default value in `config`.

vers-w commented on Wflow/test/bmi.jl:
```diff
@@ -176,8 +176,7 @@ end
     satwaterdepth = mean(model.land.soil.variables.satwaterdepth)
-    model.config.model.cold_start__flag = false
-    Wflow.load_state(model)
+    model = BMI.initialize(Wflow.Model, tomlpath; model_cold_start__flag = false)
```
Is this also possible with using `Wflow.load_state(model)`? This is for example important for OpenDA coupling (without initializing the `model` again).

vers-w replied:
Ah, OK, yes, for this to work (setting states in the model without re-initializing) the `model.cold_start__flag` needs to be mutable.

vers-w reviewed (changes requested):
Nice work @SouthEndMusic ! 

Besides a couple of comments/questions, the code coverage has decreased and is failing, also at the project level, would be good to bump this up. I think it would be good if @JoostBuitink also checks the config settings in `config.jl`.

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,221 @@
+end
+
+# Model configurations
+@option struct ModelSection <: AbstractConfigSection
```
That would be nice, but then we have to either:

- Also require this in the TOML, which is breaking
- Do some more processing of the TOML read before passing it to `Configurations.jl`, which sort of defeats the purpose of `Configurations.jl`

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,221 @@
+    basin__local_drain_direction::String
+    basin_pit_location__mask::Union{Nothing, String} = nothing
+    river_location__mask::String
+    reservoir_area__count::String = ""
+    reservoir_location__count::String = ""
```
You mean as a comment here? I don't think `Configurations.jl` supports validation logic like 'if this condition for this field, then that condition for that field'. I think it's best if such conditional validation is done in the code where the field is parsed.

vers-w replied:
OK, yeah, it seems indeed `Configurations.j` does not support this kind of validation. But it would be nice to add this validation to the code, if `reservoir__flag` is `true`, `reservoir_area__count` and `reservoir_location__count` should be set (not default empty string).

SouthEndMusic replied:
OK, I'll set the default to `nothing` and then raise an error when the data is required.

JoostBuitink replied:
I think it would be best to validate this already when reading the config. Given that Configurations.jl does not support this, and appears to be not too actively maintained, I wonder if it is worth writing our own logic around this, or whether it is better to define our own config object. 

It would also be nice to check if certain layers required specifically to the model type/settings (for example layers like soil moisture, glacier locations, reservoir parameters, floodplain volumes) are present. It is a bit tricky that those can be defined both in the `[input.static]` or `[input.cyclic]` settings.

Also, some fields like `river_routing` only support two options ("kinematic-wave" or "local-inertial"). If a user tries something else (or a typo), it would be best to throw an error when creating the config object (rather than within the computation code/initialization like it is now). I once tried to get this sorted using Configurations, but failed to do so.

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,221 @@
+    reservoir_location__count::String = ""
+    subbasin_location__count::String
+    # Ouput locations: these are not directly part of the model
+    river_gauge__count::String = ""
+    river_gauge_grdc__count::String = ""
```
Good to know. This is never accessed in the Wflow core right? Currently `Configurations.jl` is very strict, in that if a TOML section is parsed with an `@option` like this, every non-hardcoded field yields an error. I made an issue: https://github.com/Roger-luo/Configurations.jl/issues/107. The package isn't actively maintained, the last release was 2 years ago. If there is no response to my issue I can take a stab at it myself.

vers-w replied:
> Good to know. This is never accessed in the Wflow core right? Currently `Configurations.jl` is very strict, in that if a TOML section is parsed with an `@option` like this, every non-hardcoded field yields an error. I made an issue: [Roger-luo/Configurations.jl#107](https://github.com/Roger-luo/Configurations.jl/issues/107). The package isn't actively maintained, the last release was 2 years ago. If there is no response to my issue I can take a stab at it myself.

Access in Wflow core is done here: [io.jl L650](https://github.com/Deltares/Wflow.jl/blob/toml_parse/Wflow/src/io.jl#L650), is that what you mean? These fields are not hard-coded. Hm, if the package is not actively maintained, this seems a bit risky to me. I would say if we could implement something similar without too much extra work, that would be the preferred option I think. Also good to know your thoughts on this @JoostBuitink (as you did try out this package for wflow a while ago)?

JoostBuitink replied:
I think that would be nice, as the input section is indeed extremely flexible, where users can define a variable with any name and map it to a layer in the netcdf (staticmaps.nc). But indeed, there are a couple of fields mandatory to be defined. 

Your (@SouthEndMusic) sounds good to me! I just have the slight concern that we are writing quite some logic around a a package that is not actively maintained, and wonder if creating our own code around this is actually easier..

SouthEndMusic replied:
@JoostBuitink I see what you mean. For Ribasim we've been using `Configurations.jl` for a while, so we're used to working with its constraints. The bottom line is that it's hard to fit the very flexible Wflow TOML into a schema. Indeed, if we need more and more workarounds we should make something custom, but I think for what I have now it's still worth using.

vers-w replied:
@SouthEndMusic and @JoostBuitink : suggest to discuss this tomorrow further; what is the added value of using `Configurations.jl` while it is not actively maintained and considering the constraints? How much extra work is needed for a custom implementation?

SouthEndMusic reviewed (commented):
/

vers-w commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,314 @@
+# Logging related configurations
+@option struct LoggingSection <: AbstractConfigSection
+    silent::Bool = false
+    loglevel::String = "info"
```
From checking the code in logging.jl (and the docs) this is allowed.

vers-w commented on Wflow/src/io.jl:
```diff
@@ -773,16 +663,13 @@ function nc_variables_dims(nc_variables, dataset, config)
+                    var,
                     location_dim = nc_var["location"],
                     locations = [nc_var["location"]],
                 ),
```
This is not vaildated somewhere else. What kind of validation were you thinking of?

SouthEndMusic commented:
After discussing with @evetion I'm motivated to keep `Configurations.jl` as a dependency, because:

- The only part of the configuration that has both standard and flexible fields is the input section. This is handled in a simple wrapper of the default `Configurations.from_dict` for this section.
- Other parts fit nicely in a schema as defined by `Configurations.jl`, including the very flexible output section. This means that a lot of parsing of the entries here can be done (in a simpler way) in `config.jl` than in `io.jl`.
- In general, overloading of `Configurations.from_dict` avoids all needs to work around `Configurations.jl` as I did before. This is within the documented usage of `Configurations.jl`, specifically [this part](https://configurations.rogerluo.dev/stable/convert/).
- A particularly nice conversion that we can let `Configurations.jl` do this way is from string options to enumerator (`EnumX.Enum`) instances. This means that we have a centralized way to define which options a particular part of the configuration has and to throw an informative error if an invalid option was given in the TOML.

My apologies for making the PR so big. I do believe that (most of) this is a good step in making the code easier to understand as a newcomer

vers-w commented:
> After discussing with @evetion I'm motivated to keep `Configurations.jl` as a dependency, because:

I am a bit puzzled, my impression is that, as we also did discuss last week, that there are not that many structs that we could initialize in a customized way without using `Configurations.jl` as a dependency. With the `@with_kw` macro, default values are possible, and I think initializing these structs from the parsed TOML file (`TOML.parsefile`) can be done quite easily. This would also give us more flexibility, for example related to https://github.com/Deltares/Wflow.jl/pull/689#discussion_r2336058621, and for any future changes we do not have to deal with some of the `Configurations.jl` restrictions. Finally, it is probably also easier to understand and to expand on by a newcomer.

JoostBuitink commented on Wflow/test/run.jl:
```diff
@@ -5,58 +5,72 @@ using NCDatasets
 model = Wflow.Model(config)
 Wflow.run!(model)
 
 # second half of January, warm start, fews_run set to true, and starttime set one day earlier
 # to match endtime of part 1
-config.time.starttime = DateTime("2000-01-14T00:00:00")
-config.time.endtime = DateTime("2000-02-01T00:00:00")
-config.model.cold_start__flag = false  # warm start
-config.fews_run__flag = true
-config.state.path_input =
-    joinpath(dirname(tomlpath), "data/state-test/outstates-moselle-january-1of2.nc")
-config.state.path_output = joinpath(
-    dirname(tomlpath),
-    "data/state-test/outstates-moselle-january-2of2-fews_run.nc",
+config = Wflow.Config(
+    tomlpath;
+    time_starttime = DateTime("2000-01-14T00:00:00"),
+    time_endtime = DateTime("2000-02-01T00:00:00"),
+    model_cold_start__flag = false, # warm start
+    fews_run__flag = true,
+    state_path_input = joinpath(
+        dirname(tomlpath),
+        "data/state-test/outstates-moselle-january-1of2.nc",
+    ),
+    state_path_output = joinpath(
+        dirname(tomlpath),
+        "data/state-test/outstates-moselle-january-2of2-fews_run.nc",
+    ),
```
Still a bit in doubt whether I like this new implementation over the old one. To me, it feels a bit more natural to read the toml in one step, and then allow for modifications afterwards. Now these two have be done in the same "step". 

Not a massive change, or maybe just something that I need to get used to. Although changing specific fields (e.g. `time_starttime`) is maybe less intuitive than `config.time.starttime = ...` )

SouthEndMusic replied:
The way this works is now a bit different because I don't use `Configurations.jl` anymore: the override of the values in the TOML is passed as a dict. My biased opinion is that config modification should be discouraged because it makes the code harder to understand. Re-initializing the config instead if modifying it is one of the steps to make the tests more self contained which is preferred by the Julia VSCode plugin. Although I don't know whether in-place config modification is needed outside the context of the tests (I know it is done in a few places in the core code, I left the relevant fields mutable).

JoostBuitink commented on Wflow/test/run_sbm.jl:
```diff
@@ -155,25 +155,22 @@ Wflow.close_files(model; delete_output = false)
-config = Wflow.Config(tomlpath)
-config.model.snow_gravitional_transport__flag = false
 
 model = Wflow.Model(config)
 Wflow.run_timestep!(model)
 
 @testset "lateral snow transport off" begin
     snow = model.land.snow
-    @test snow.variables.snow_storage[5] ≈ 3.7686103651001375
+    @test snow.variables.snow_storage[5] ≈ 3.768513390588815
```
should the toml be read again and the flag explicitly set to false? Because I think that is causing the snow_storage to be different (`snow_gravitional_transport__flag` is set to true for this test I think).

SouthEndMusic replied:
I have a few failing tests still with the latest implementation, let's discuss together.

JoostBuitink commented on Wflow/test/run_sbm_gwf.jl:
```diff
@@ -110,15 +111,20 @@ Wflow.close_files(model; delete_output = false)
+config = Wflow.Config(
+    tomlpath;
+    model_river_routing = "local-inertial",
+    model_land_routing = "local-inertial",
+)
+
+config.input.static.dict["river_bank_water__elevation"] =
+    Wflow.InputEntry(; standard_name = "bankfull_elevation")
+config.input.static.dict["river_bank_water__depth"] =
+    Wflow.InputEntry(; standard_name = "bankfull_depth")
+config.input.static.dict["land_surface_water_flow__ground_elevation"] =
+    Wflow.InputEntry(; standard_name = "wflow_dem")
```
Related to my earlier comment, here I am also doubting whether this new implementation is actually easier. To me, the previous option is much easier and more straightforward. Similar as before, only expert users will touch this interface, but still a remark I wanted to make

JoostBuitink commented on docs/model_docs/model_settings.qmd:
```diff
@@ -58,7 +58,6 @@ The Tables in this subsection show the available models settings for the model t
 | `land_surface_water_flow_threshold__depth` | Water depth threshold for calculating overland flow between cells | m | 0.001 |
 | `river_water_flow__froude_limit_flag` | Limit river flow to subcritical-critical according to Froude number | - | `true` |
 | `land_surface_water_flow__froude_limit_flag` | Limit overland flow to subcritical-critical according to Froude number | - | `true` |
-| `river_water_flow__froude_limit_flag` | Limit river flow to subcritical-critical according to Froude number | - | `true` |
```
It is actually listed twice in this table ;)

JoostBuitink reviewed (commented):
Nice work, and massive PR indeed ;) Thanks a lot for all the other code-improvements/small refactors!

I have added a couple of remarks/questions/suggestions, I hope they are all clear. Overall - but although it might be a bit late given that all this work is already there - I have the same doubts as Willem on whether the (sort of) "rigidness" of Configurations.jl and the "flexibility" of the Wflow.jl toml are a good match. Let's continue the discussion tomorrow!

SouthEndMusic commented on Wflow/src/bmi.jl:
```diff
@@ -369,7 +364,7 @@ function save_state(model::Model)
 
 function get_start_unix_time(model::Model)
-    return datetime2unix(DateTime(model.config.time.starttime))
+    return datetime2unix(model.config.time.starttime)
```
I did a quick check and this works now with the new implementation without `Configurations.jl`. This would have been fixable with `Configurations.jl` though

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,581 @@
+@enumx ModelType sbm sbm_gwf sediment
+@enumx CalendarType standard gregorian proleptic_gregorian julian noleap _365_day all_leap _366_day _360_day
+@enumx ConductivityProfileType uniform exponential
+@enumx SHCPType exponential exponential_constant layered layered_exponential
```
Fair point, these are quite out of context here. Although, the code base is full of abbreviations that you are used to but (non-hydrological) newcomers are not 😉

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,581 @@
+    netcdf_variable_name::Union{Nothing, String} = nothing # Comes from dict["netcdf"]["variable"]["name"]
+    scale::Vector{Float64} = [1.0]
+    offset::Vector{Float64} = [0.0]
+    layer::Union{Nothing, Vector{Int}} = nothing
```
A scalar entry here is still possible, this is converted to a vector to have a consistent type. It's worth checking whether this is parsed properly lateron.

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,581 @@
+    # Option 2
+    value::Any = nothing
+    # Option 3
+    standard_name::Union{Nothing, String} = nothing
```
I came to this implementation from looking at the different ways in which this part of the config was parsed before. If different field names here are more clear, this can be adjusted. It must still be clear in what way this part of the config was initialized, and that might be lost when `netcdf_variable_name` and `standard_name` are put in the same field. Let's discuss this together.

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,581 @@
+
+@option struct InputEntry <: AbstractConfigSection
+    # Option 1
+    netcdf_variable_name::Union{Nothing, String} = nothing # Comes from dict["netcdf"]["variable"]["name"]
```
Yes, I wondered why this was nested so deeply but I figured there would be a reason for it. I agree that the new way is cleaner, but it is breaking.

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,581 @@
+
+@option struct CoordinateSection <: AbstractConfigSection
+    x::Float64
+    y::Float64
```
I don't think I can answer this question directly, let's discuss.

SouthEndMusic commented on Wflow/src/config.jl:
```diff
@@ -0,0 +1,581 @@
+    index::Union{Nothing, Int, IndexSection} = nothing
+    map::Union{Nothing, String} = nothing
+    coordinate::Union{Nothing, CoordinateSection} = nothing
+    location::Union{Nothing, String} = nothing
```
An interesting consequence of making the config structure explicit 😄

SouthEndMusic commented on Wflow/src/io.jl:
```diff
@@ -184,10 +26,11 @@ function get_at(
+function get_at(ds::CFDataset, var::InputEntry, i)
+    (; scale, offset) = var
+    data = read_standardized(ds, variable_name(var), (x = :, y = :, time = i))
+    if scale != 1.0 || offset != 0.0
+        data .= data .* scale .+ offset
```
That is a good point, I also wondered this. Is the computation expensive enough to warrant this condition? Let's discuss.

SouthEndMusic commented:
@JoostBuitink @vers-w I have addressed your comments. Most of the old interface for mutating the config is now intact, type conversions are handled automatically. One thing that I think is worth reviewing is what the info messages now look like. The order has changed, and some things are also formatted differently.

vers-w commented:
Just did a final check of the logging messages:

- Not sure if it is required to log when a parameters is modified with scale 1.0 and offset 0.0:
  `Info: NetCDF parameter wflow_ldd is modified with scale 1.0 and offset 0.0.`
   I think it is better to remove these, @JoostBuitink ?
- It would be nice to use single quotes for the parameter names and values (this is clearer in the console) and already used. 
- Something goes wrong with formatting this external parameter name?:
  `[ Info: NetCDF parameter wflowgaugesgrdc is modified with scale 1.0 and offset 0.0.`

SouthEndMusic commented:
@vers-w I resolved all the merge conflicts but there is still one test failing that I don't know how to fix. Can you have a look?

vers-w commented:
> @vers-w I resolved all the merge conflicts but there is still one test failing that I don't know how to fix. Can you have a look?

@SouthEndMusic : I have resolved the tests and changed the log message for modifying NetCDF parameters (https://github.com/Deltares/Wflow.jl/pull/689/commits/5bcfa41bf1cfd4d2e8617b5041990bbe7a30783e). I think it is fine to merge, @JoostBuitink: maybe you want to do a final check?
